pwiz_tools/Skyline/FileUI/SkypSupport.cs

```diff
@@ -32,111 +27,284 @@ public SkypSupport(SkylineWindow skyline)
+                        var exception = progressStaus.ErrorException;
+                        var skypEx = exception as SkypDownloadException;
+                        if (skypEx != null)
+                        {
+                            if (skypEx.Unauthorized())
+                            {
+                                return skyp.Server == null ? 
+                                    AddServerAndOpen(skyp, skypEx.Message, parentWindow)  // Server not saved in Skyline. Offer to add the server.
+                                    : EditServerAndOpen(skyp, skypEx.Message, parentWindow); // Server saved in Skyline but credentials are invalid. Offer to edit server credentials.
+                            }
+                            else if (skypEx.Forbidden() && skyp.UsernameMismatch())
+                            {
+                                // Server is saved in Skyline but the user in the saved credentials does not have enough permissions. 
+                                // The downloading user in the skyp file is different from saved user.  Offer to edit server credentials. 
+                                return EditServerAndOpen(skyp, skypEx.Message, parentWindow);
+                            }
+                            
+                            MessageDlg.ShowWithException(parentWindow ?? _skyline, skypEx.Message, skypEx);
+                            return false;
+                        }
+                        else
+                        {
+                            ShowDownloadError(parentWindow, skyp, exception);
+                            return false;
+                        }
+                    }
                 }
+                
                 return _skyline.OpenSharedFile(skyp.DownloadPath);
             }
             catch (Exception e)
             {
-                if (_retryRemaining-- > 0 && skyp != null && e.Message.Contains(AddPanoramaServerMessage(skyp)))
-                {
-                    return AddServerAndOpen(skypPath, e.Message, parentWindow);
-                }
-                else
+                ShowDownloadError(parentWindow, skyp, e);
+                return false;
+            }
+        }
+
+        private void ShowDownloadError(FormEx parentWindow, SkypFile skyp, Exception exception)
+        {
+            var message = TextUtil.LineSeparate(string.Format(
+                Resources
+                    .SkypSupport_Download_There_was_an_error_downloading_the_Skyline_document_specified_in_the_skyp_file___0__,
+                skyp.SkylineDocUri), exception.Message);
+            MessageDlg.ShowWithException(parentWindow ?? _skyline, message, exception);
+        }
+
+        private bool AddServerAndOpen(SkypFile skypFile, string message, FormEx parentWindow)
+        {
+            using (var alertDlg = new AlertDlg(message, MessageBoxButtons.OKCancel))
+            {
+                alertDlg.ShowDialog(parentWindow ?? _skyline);
```

Comment:
Just a nitpick but you can check the return value of ShowDialog instead of alertDlg.DialogResult, e.g.
```csharp
if (alertDlg.ShowDialog(parentWindow ?? _skyline) != DialogResult.OK)
{
    return false;
}
```